integration/framework/service.go

```diff
@@ -0,0 +1,330 @@
+package framework
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/pkg/errors"
+	"github.com/prometheus/common/expfmt"
+)
+
+var (
+	dockerPortPattern = regexp.MustCompile(`^.*:(\d+)$`)
+)
+
+type Service struct {
+	name         string
+	image        string
+	networkName  string
+	networkPorts []int
+	env          map[string]string
+	command      *Command
+	readiness    *ReadinessProbe
+
+	// Maps container ports to dynamically binded local ports.
+	networkPortsContainerToLocal map[int]int
+
+	// Generic retry backoff.
+	retryBackoff *util.Backoff
+}
+
+func NewService(
+	name string,
+	image string,
+	networkName string,
+	networkPorts []int,
+	env map[string]string,
+	command *Command,
+	readiness *ReadinessProbe,
+) *Service {
+	return &Service{
+		name:                         name,
+		image:                        image,
+		networkName:                  networkName,
+		networkPorts:                 networkPorts,
+		env:                          env,
+		command:                      command,
+		networkPortsContainerToLocal: map[int]int{},
+		readiness:                    readiness,
+		retryBackoff: util.NewBackoff(context.Background(), util.BackoffConfig{
+			MinBackoff: 100 * time.Millisecond,
+			MaxBackoff: 200 * time.Millisecond,
+			MaxRetries: 100, // Sometimes the CI is slow
+		}),
+	}
+}
+
+func (s *Service) Start() error {
```

Comment:
Another issue I've run into is that container already existed. Would it be possible to delete them beforehand? Or even better, use dynamic container names, but static hostnames on the network? (That's like V2 request)

Reply:
I thought about it. I would be more glad to fix the issues leading to the "container already existed". Since we're using a shared Docker environment, I would prefer to not remove containers just because of a matching name (ie. it's not guaranteed that the "dynamodb" container is actually the integration test container if we find it already running on the system).

> Or even better, use dynamic container names, but static hostnames on the network?

I thought about it, but makes things even more complicated. I'm wondering if we could keep it for the future if we see this being an issue after the fix done [in this commit](https://github.com/cortexproject/cortex/pull/1839/commits/7d7be2d5f911802cc2f06697b74abbf132245d03).

What's your take?

Reply:
I'm fine pushing dynamic container names for (possibly much) later. With latest fix, I cannot reproduce problem with unremoved containers.